accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -0,0 +1,1098 @@
+
+*For example, the extension type erasure could map
+<code>X extends C<Y>, Y extends X</code> to
+<code>X extends Y, Y extends X</code>,
+which is an error.*
+
+An extension type declaration _DV_ named `Name` may declare one or
+more constructors. A constructor which is declared in an extension type
+declaration is also known as an _extension type constructor_.
+
+*The purpose of having an extension type constructor is that it bundles
+an approach for building an instance of the representation type of an
+extension type declaration _DV_ with _DV_ itself, which makes it easy to
+recognize that this is a way to obtain a value of that extension type. It
+can also be used to verify that an existing object (provided as an
+actual argument to the constructor) satisfies the requirements for
+having that extension type.*
+
+The `<representationDeclaration>` works as a constructor. The optional
+`('.' <identifier>)` in the grammar is used to declare this constructor
+with a name of the form `<identifier> '.' <identifier>` *(at times
+described as a "named constructor")*. It is a constant constructor: If `e`
+is a constant expression and `V(e)` is not an error, then `V(e)` is a
+constant expression. Other constructors may be declared `const` or not,
+following the normal rules for constant constructors.
+
+A compile-time error occurs if an extension type constructor includes a
+superinitializer. *That is, a term of the form `super(...)` or
+`super.id(...)` as the last element of the initializer list.*
+
+A compile-time error occurs if an extension type constructor declares a
+super parameter. *For instance, `Name(super.x);`.*
+
+*In the body of a generative extension type constructor, the static type of
+`this` is the same as it is in any instance member of the extension type
+declaration, that is, `Name<X1 .. Xk>`, where `X1 .. Xk` are the type
+parameters declared by `Name`.*
+
+An instance creation expression of the form
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;(...)</code>
+or
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;.name(...)</code>
+is used to invoke these constructors, and the type of such an expression is
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>.
+
+*In short, extension type constructors appear to be very similar to
+constructors in classes, and they correspond to the situation
+where the enclosing class has a single, non-late, final instance
+variable, which is initialized according to the normal rules for
+constructors (in particular, it can occur by means of `this.id`, or in
+an initializer list, or by an initializing expression in the
+declaration itself, but it is an error if it does not occur at all).*
+
+An extension type `V` used as an expression (*a type literal*) is allowed
+and has static type `Type`.
+
+An extension type declaration _DV_ can have the type modifier
+`final`. In this case it is a compile-time error for another extension type
+declaration _DV2_ to have _DV_ as a superinterface.
```

Comment:
Is this limited to extension types outside the library declaring _DV_, or is there an intentional difference here? (If the latter, consider making the distinction explicit in the text.)

Reply:
My intention was to make it more strict, but I've changed it back to the approach where the error only occurs if _DV_ and _DV2_ are declared in different libraries. Consistency is a value on its own, in spite of the fact that it can be tempting to break it. ;-)